servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/FromNPublisher.java

```diff
@@ -17,47 +17,58 @@
 
 import javax.annotation.Nullable;
 
+import static io.servicetalk.concurrent.internal.FlowControlUtils.addWithOverflowProtection;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.handleExceptionFromOnSubscribe;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.newExceptionForInvalidRequestN;
+import static java.lang.Math.min;
 
-final class From2Publisher<T> extends AbstractSynchronousPublisher<T> {
+final class FromNPublisher<T> extends AbstractSynchronousPublisher<T> {
     @Nullable
     private final T v1;
     @Nullable
     private final T v2;
+    @Nullable
+    private final T v3;
+
+    private final int maxN;
 
-    From2Publisher(@Nullable T v1, @Nullable T v2) {
+    FromNPublisher(@Nullable T v1, @Nullable T v2) {
         this.v1 = v1;
         this.v2 = v2;
+        this.v3 = null;
+        this.maxN = 2;
+    }
+
+    FromNPublisher(@Nullable T v1, @Nullable T v2, @Nullable T v3) {
+        this.v1 = v1;
+        this.v2 = v2;
+        this.v3 = v3;
+        this.maxN = 3;
     }
 
     @Override
     void doSubscribe(final Subscriber<? super T> subscriber) {
         try {
-            subscriber.onSubscribe(new TwoValueSubscription(subscriber));
+            subscriber.onSubscribe(new NValueSubscription(subscriber));
         } catch (Throwable cause) {
             handleExceptionFromOnSubscribe(subscriber, cause);
         }
     }
 
-    private final class TwoValueSubscription implements Subscription {
-        private static final byte INIT = 0;
-        private static final byte DELIVERED_V1 = 1;
-        private static final byte CANCELLED = 2;
-        private static final byte TERMINATED = 3;
-        private byte state;
+    private final class NValueSubscription implements Subscription {
+        private static final int TERMINATED = -1;
+        private int requested;
+        private int state;
         private final Subscriber<? super T> subscriber;
 
-        private TwoValueSubscription(final Subscriber<? super T> subscriber) {
+        private NValueSubscription(final Subscriber<? super T> subscriber) {
             this.subscriber = subscriber;
         }
 
         @Override
         public void cancel() {
-            if (state != TERMINATED) {
```

Comment:
iirc this check was added so that terminal state could be propagated regardless of cancellation or not (e.g. an error).

Reply:
this is only valid for handling invalid `request(n)`. the spec says that after cancelation it should be [nop](https://github.com/reactive-streams/reactive-streams-jvm#3.6). validation of `N` is not required at this point. hence, better to minimize number of states

Reply:
I believe @Scottmitch is referring to this case:

```
void testErrorWhenCancelInterleavesDemandCompletion() {
        final int[] emitted = {0};
        final boolean[] completed = {false};
        final boolean[] error = {false};
        final Subscription[] subscription = new Subscription[1];
        toSource(fromPublisher()).subscribe(new Subscriber<Integer>() {
            @Override
            public void onSubscribe(final Subscription subscription1) {
                subscription[0] = subscription1;
                subscription1.request(3);
            }

            @Override
            public void onNext(@Nullable final Integer integer) {
                subscription[0].cancel();
                if (requireNonNull(integer) == 2) {
                    throw new IllegalStateException();
                } else {
                    emitted[0]++;
                }
            }

            @Override
            public void onError(final Throwable t) {
                error[0] = true;
            }

            @Override
            public void onComplete() {
                completed[0] = true;
            }
        });

        assertThat(emitted[0], is(1));
        assertThat(completed[0], is(false));
        assertThat(error[0], is(true));
    }
```

Allowing the terminal signal to be propagated, especially when this can happen async. The spec says that the `cancel` could be eventual. I think the spec allows both behaviours, but allowing an error to be delivered I think is preferable. 
Thoughts?